### PR TITLE
mgr/cephadm: limit number of times check host is performed in the serve loop

### PR DESCRIPTION
**Pull request by Daniel-Pivonka**
limit number of times check host is performed in the serve loop

we only need this to happen every 10 min on hosts

Fixes: https://tracker.ceph.com/issues/44401

Signed-off-by: Daniel-Pivonka <dpivonka@redhat.com>

### CODE REVIEW

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -989,6 +1015,28 @@ def _check_hosts(self):
             }
         self.set_health_checks(self.health_checks)
 
+    # TODO func "check_host" already exist need a name for this used __ for now
```
Comment (liewegas) [inline src/pybind/mgr/cephadm/module.py]: drop this comment

File: src/pybind/mgr/cephadm/module.py [hunk 2]
```diff
@@ -989,6 +1015,28 @@ def _check_hosts(self):
...
+    def _check_host(self, host):
+
```
Comment (liewegas) [inline src/pybind/mgr/cephadm/module.py]: and this blank line

File: src/pybind/mgr/cephadm/module.py [hunk 3]
```diff
@@ -1062,6 +1114,18 @@ def serve(self):
...
+            if 'CEPHADM_HOST_CHECK_FAILED' in self.health_checks:
+                del self.health_checks['CEPHADM_HOST_CHECK_FAILED']
+            if bad_hosts:
+                self.health_checks['CEPHADM_HOST_CHECK_FAILED'] = {
+                    'severity': 'warning',
+                    'summary': '%d hosts fail cephadm check' % len(bad_hosts),
+                    'count': len(bad_hosts),
+                    'detail': bad_hosts,
+                }
+            self.set_health_checks(self.health_checks)
```
Comment (liewegas) [inline src/pybind/mgr/cephadm/module.py]: this should be indented, and we need it in the previous if too when we delete the item.

or, make a health_changed = False, and set it to True in each of these blocks (below, too!) and do one final set_health_checks call at the bottom if health_changed...

File: src/pybind/mgr/cephadm/module.py [hunk 4]
```diff
@@ -357,6 +368,14 @@ def host_needs_device_refresh(self, host):
             return True
         return False
 
+    def host_needs_check(self, host):
+        # type: (str) -> bool
+        cutoff = datetime.datetime.utcnow() - datetime.timedelta(
+            seconds=self.mgr.host_check_timeout)
+        if host not in self.last_host_check or self.last_host_check[host] < cutoff:
+            return True
+        return False
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py]:
```suggestion
        return host not in self.last_host_check or self.last_host_check[host] < cutoff
```
?

**Comment by sebastian-philipp** [comment]
regarding the newly introduced code duplication:

image

* `_check_host` doesn't set `--expect-hostname`. why?
* `_check_host` doesn't set `addr=addr`. why?
* `check_host` doesn't log anything
* `check_host` doesn't update `self.cache`. why?
* `check_host` doesn't honor `self.warn_on_failed_host_check`. why?

**Comment by liewegas** [comment]
> regarding the newly introduced code duplication:
> 
> image
> 
>     * `_check_host` doesn't set `--expect-hostname`. why?
> 
>     * `_check_host` doesn't set `addr=addr`. why?

_run_cephadm does both of these implicitly from get_connection since it's a registered host.

>     * `check_host` doesn't log anything
> 
>     * `check_host` doesn't update `self.cache`. why?
> 
>     * `check_host` doesn't honor `self.warn_on_failed_host_check`. why?

the check_host was meant to be an explicit command that is unrelated to whether the host is registered or not, so that you could test something before adding it. we could make it do some of these things, but they would all have conditional checks for whether the host is in the inventory.. i'm not sure it makes sense.

**Comment by liewegas** [comment]
retest this please

File: src/pybind/mgr/cephadm/module.py [hunk 5]
```diff
@@ -268,6 +276,7 @@ def save_host(self, host):
             'daemons': {},
             'devices': [],
             'daemon_config_deps': {},
+            'last_host_check' : {},
```
Comment (liewegas) [inline src/pybind/mgr/cephadm/module.py]: don't need this line

File: src/pybind/mgr/cephadm/module.py [hunk 6]
```diff
@@ -558,6 +575,12 @@ class CephadmOrchestrator(orchestrator.Orchestrator, MgrModule):
             'default': 10 * 60,
             'desc': 'seconds to cache service (daemon) inventory',
         },
+        {
+            'name': 'host_check_timeout',
```
Comment (liewegas) [inline src/pybind/mgr/cephadm/module.py]: can we do ``host_check_interval`` instead?

File: src/pybind/mgr/cephadm/module.py [hunk 7]
```diff
@@ -558,6 +575,12 @@ class CephadmOrchestrator(orchestrator.Orchestrator, MgrModule):
...
+            'type': 'secs',
+            'default': 10 * 60,
+            'desc': 'seconds to cache host check',
```
Comment (liewegas) [inline src/pybind/mgr/cephadm/module.py]: how frequently to perform a host check

**Comment by liewegas** [comment]
Can you rebase this on top of the octopus branch?  ``git rebase HEAD^ --onto origin/octopus``

**Comment by liewegas** [comment]
sorry, this needs one more rebase

**Pull request merged**